packages/services/src/lib/circuit-breaker.ts

```diff
@@ -0,0 +1,372 @@
+    }
+
+    this.failureCount++;
+    this.lastFailureTime = Date.now();
+    this.successCount = 0; // Reset success count on any failure
+
+    if (this.state === "half-open") {
+      // Any failure in half-open state reopens the circuit
+      this.transition("open");
+    } else if (this.state === "closed") {
+      // Check if we should open based on threshold strategy
+      if (this.shouldOpen()) {
+        this.transition("open");
+      }
+    }
+  }
+
+  /**
+   * Record request in rolling window
+   */
+  private recordRequest(success: boolean): void {
+    const now = Date.now();
+    this.requestHistory.push({ timestamp: now, success });
+
+    // Clean old entries
+    if (this.config.rollingWindowDuration) {
+      const windowStart = now - this.config.rollingWindowDuration;
+      while (
+        this.requestHistory.length > 0 &&
+        this.requestHistory[0]!.timestamp <= windowStart
+      ) {
+        this.requestHistory.shift();
+      }
+    }
+  }
+
+  /**
+   * Determine if circuit should open based on configured strategy
+   */
+  private shouldOpen(): boolean {
+    // If using rolling window with percentage threshold
+    if (
+      this.config.rollingWindowDuration &&
+      this.config.errorThresholdPercentage
+    ) {
+      const stats = this.getWindowStats();
+      return (
+        (stats.errorRatePercentage || 0) >=
+          this.config.errorThresholdPercentage &&
+        (stats.requestsInWindow || 0) >= 5
+      ); // Minimum requests to avoid false positives
+    }
+
+    // Otherwise use consecutive failure count
+    return this.failureCount >= this.config.failureThreshold;
+  }
+
+  /**
+   * Transition between states
+   */
+  private transition(to: CircuitState): void {
+    const from = this.state;
+    this.state = to;
+
+    // Reset counters on state transition
+    if (to === "closed") {
+      this.failureCount = 0;
+      this.successCount = 0;
+    } else if (to === "half-open") {
+      this.successCount = 0;
+    }
+
+    // Log state change
+    this.config.logger?.info(
+      {
+        circuitBreaker: this.config.name,
+        from,
+        to,
+        failureCount: this.failureCount,
+      },
+      `[CircuitBreaker] State transition: ${from} -> ${to}`,
+    );
+
+    // Call optional callback
+    this.config.onStateChange?.(from, to);
+  }
+}
+
+/**
+ * Factory function to create a circuit breaker with defaults for financial/trading APIs
+ * Based on industry best practices research
+ */
+export function createCircuitBreaker(
+  name: string,
+  options: Partial<CircuitBreakerConfig> = {},
+): CircuitBreaker {
+  return new CircuitBreaker({
+    name,
+    failureThreshold: 3, // Lower threshold for financial APIs
+    resetTimeout: 10000, // 10 seconds (faster recovery for trading systems)
+    successThreshold: 2, // Require 2 successful calls to fully close
+    requestTimeout: 3000, // 3 second timeout for API calls
+    ...options,
+  });
+}
```

Comment:
Should we keep/use this mode at all? Using this could result in the same issue we identified in a previous PR - a single success can reset a streak of failures.

Since you already have the rolling-window circuit breaker, should we use that instead?

Reply:
(But otherwise, this is great!!)